File: backend/app/modules/management_governance/overview.py
"""Overview summary — ports frontend overview.ts."""

from __future__ import annotations

from typing import Any

from app.modules.management_governance.assessment import build_governance_assessment
from app.modules.management_governance.compute import compute_management_governance_model
from app.modules.management_governance.constants import SECTION_LABELS
from app.modules.management_governance.progress import calculate_management_governance_progress


def build_overview_summary(
    payload: dict[str, Any],
    linked_references: dict[str, Any] | None = None,
) -> dict[str, Any]:
    linked = linked_references or {}
    progress = calculate_management_governance_progress(payload)
    model = compute_management_governance_model(payload, linked)
    assessment = build_governance_assessment(payload, model, progress, linked)

    sections_in_progress = sum(
        1 for status in progress["sections"].values() if status == "in_progress"
    )

    incomplete_sections = [
        (section_id, status)
        for section_id, status in progress["sections"].items()
        if status != "complete"
    ]

    recommended_next_actions = [
        {
            "sectionId": section_id,
            "label": f"Continue with {SECTION_LABELS.get(section_id, section_id)}",
        }
        for section_id, _ in incomplete_sections[:4]
    ]

    continuity = model.get("continuity") or {}
    board_counts = model.get("boardCounts") or {}
    applicability = model.get("applicability") or {}

    return {
        "sectionStatuses": progress["sections"],
        "sectionsComplete": progress["sectionsComplete"],
        "sectionsInProgress": sections_in_progress,
        "totalSections": progress["totalSections"],
        "overallStatus": progress["overallStatus"],
        "boardSize": model.get("boardSize", 0),
        "proposedBoardSize": model.get("proposedBoardSize", 0),
        "executiveDirectors": board_counts.get("executive", 0),
        "nonExecutiveDirectors": board_counts.get("nonExecutive", 0),
        "independentDirectors": board_counts.get("independent", 0),
        "womenDirectors": board_counts.get("women", 0),
        "residentDirectors": board_counts.get("resident", 0),
        "chairmanName": model.get("chairmanName", ""),
        "managingDirectorName": model.get("managingDirectorName", ""),
        "kmpCount": model.get("kmpCount", 0),
        "seniorManagementCount": model.get("smpCount", 0),
        "criticalVacancies": continuity.get("criticalRoleVacancies", 0),
        "committeesReady": model.get("committeesReadyCount", 0),
        "committeesRequired": model.get("committeesRequiredCount", 0),
        "policiesAdopted": model.get("policiesAdoptedCount", 0),
        "policiesRequired": model.get("policiesRequiredCount", 0),
        "boardChangesLastThreeYears": continuity.get("boardAdditionsLastThreeYears", 0)
        + continuity.get("boardCessationsLastThreeYears", 0),
        "kmpChangesLastThreeYears": continuity.get("kmpSmpAdditionsLastThreeYears", 0)
        + continuity.get("kmpSmpCessationsLastThreeYears", 0),
        "pendingAppointments": model.get("pendingAppointments", 0),
        "potentialConcerns": assessment["counts"]["potentialConcern"],
        "professionalReviewItems": assessment["counts"]["pendingProfessionalConfirmation"],
        "listingSegment": applicability.get("listingSegment", "unknown"),
        "assessmentResult": assessment["result"],
        "assessmentResultLabel": assessment["resultLabel"],
        "assessmentSummary": assessment["summary"],
        "recommendedNextActions": recommended_next_actions,
    }

File: frontend/lib/management-governance/overview.ts
/**
 * Overview summary derived from the in-memory Management & Governance draft (M1).
 */

import { buildGovernanceAssessment, type GovernanceAssessmentResponse } from '@/lib/management-governance/assessment';
import {
  computeManagementGovernanceModel,
  type ManagementGovernanceModel,
} from '@/lib/management-governance/compute';
import { calculateManagementGovernanceProgress } from '@/lib/management-governance/progress';
import {
  createEmptyLinkedWorkstreamReferences,
  type LinkedWorkstreamReferences,
  type ManagementGovernanceProgress,
} from '@/lib/management-governance/types';
import type {
  ManagementGovernancePayload,
  ManagementGovernanceSectionId,
} from '@/lib/schemas/management-governance';
import { MANAGEMENT_GOVERNANCE_SECTION_LABELS } from '@/lib/management-governance/options';

export type ManagementGovernanceOverviewSummary = {
  sectionStatuses: ManagementGovernanceProgress['sections'];
  sectionsComplete: number;
  sectionsInProgress: number;
  totalSections: number;
  overallStatus: ManagementGovernanceProgress['overallStatus'];
  boardSize: number;
  proposedBoardSize: number;
  executiveDirectors: number;
  nonExecutiveDirectors: number;
  independentDirectors: number;
  womenDirectors: number;
  residentDirectors: number;
  chairmanName: string;
  managingDirectorName: string;
  kmpCount: number;
  seniorManagementCount: number;
  criticalVacancies: number;
  committeesReady: number;
  committeesRequired: number;
  policiesAdopted: number;
  policiesRequired: number;
  boardChangesLastThreeYears: number;
  kmpChangesLastThreeYears: number;
  pendingAppointments: number;
  potentialConcerns: number;
  professionalReviewItems: number;
  listingSegment: string;
  assessmentResult: GovernanceAssessmentResponse['result'];
  assessmentResultLabel: string;
  assessmentSummary: string;
  recommendedNextActions: Array<{ sectionId: ManagementGovernanceSectionId; label: string }>;
};

export function buildOverviewSummary(
  payload: ManagementGovernancePayload,
  linkedReferences: LinkedWorkstreamReferences = createEmptyLinkedWorkstreamReferences(),
): ManagementGovernanceOverviewSummary {
  const progress = calculateManagementGovernanceProgress(payload);
  const model = computeManagementGovernanceModel(payload, linkedReferences);
  const assessment = buildGovernanceAssessment(payload, model, progress, linkedReferences);

  const sectionsInProgress = Object.values(progress.sections).filter(
    (status) => status === 'in_progress',
  ).length;

  const incompleteSections = (
    Object.entries(progress.sections) as Array<
      [ManagementGovernanceSectionId, ManagementGovernanceProgress['sections'][ManagementGovernanceSectionId]]
    >
  ).filter(([, status]) => status !== 'complete');

  const recommendedNextActions = incompleteSections.slice(0, 4).map(([sectionId]) => ({
    sectionId,
    label: `Continue with ${MANAGEMENT_GOVERNANCE_SECTION_LABELS[sectionId]}`,
  }));

  return {
    sectionStatuses: progress.sections,
    sectionsComplete: progress.sectionsComplete,
    sectionsInProgress,
    totalSections: progress.totalSections,
    overallStatus: progress.overallStatus,
    boardSize: model.boardSize,
    proposedBoardSize: model.proposedBoardSize,
    executiveDirectors: model.boardCounts.executive,
    nonExecutiveDirectors: model.boardCounts.nonExecutive,
    independentDirectors: model.boardCounts.independent,
    womenDirectors: model.boardCounts.women,
    residentDirectors: model.boardCounts.resident,
    chairmanName: model.chairmanName,
    managingDirectorName: model.managingDirectorName,
    kmpCount: model.kmpCount,
    seniorManagementCount: model.smpCount,
    criticalVacancies: model.continuity.criticalRoleVacancies,
    committeesReady: model.committeesReadyCount,
    committeesRequired: model.committeesRequiredCount,
    policiesAdopted: model.policiesAdoptedCount,
    policiesRequired: model.policiesRequiredCount,
    boardChangesLastThreeYears:
      model.continuity.boardAdditionsLastThreeYears + model.continuity.boardCessationsLastThreeYears,
    kmpChangesLastThreeYears:
      model.continuity.kmpSmpAdditionsLastThreeYears + model.continuity.kmpSmpCessationsLastThreeYears,
    pendingAppointments: model.pendingAppointments,
    potentialConcerns: assessment.counts.potentialConcern,
    professionalReviewItems: assessment.counts.pendingProfessionalConfirmation,
    listingSegment: model.applicability.listingSegment,
    assessmentResult: assessment.result,
    assessmentResultLabel: assessment.resultLabel,
    assessmentSummary: assessment.summary,
    recommendedNextActions,
  };
}

/** Alias used by barrels and external imports. */
export const buildManagementGovernanceOverviewSummary = buildOverviewSummary;
